network-libp2p/src/dispatch/codecs/typed.rs

```diff
@@ -27,30 +23,106 @@ pub struct MessageCodec {}
+    let mut len_data = unsigned_varint::encode::usize_buffer();
+    let encoded_len = unsigned_varint::encode::usize(len, &mut len_data).len();
+    socket.write_all(&len_data[..encoded_len]).await?;
+
+    Ok(())
+}
+
+/// Reads a variable-length integer from the `socket`.
+///
+/// As a special exception, if the `socket` is empty and EOFs right at the beginning, then we
+/// return `Ok(0)`.
+///
+/// > **Note**: This function reads bytes one by one from the `socket`. It is therefore encouraged
+/// >           to use some sort of buffering mechanism.
+async fn read_varint(socket: &mut (impl AsyncRead + Unpin)) -> Result<usize, io::Error> {
+    let mut buffer = unsigned_varint::encode::usize_buffer();
+    let mut buffer_len = 0;
+
+    loop {
+        match socket.read(&mut buffer[buffer_len..buffer_len + 1]).await? {
+            0 => {
+                // Reaching EOF before finishing to read the length is an error, unless the EOF is
+                // at the very beginning of the substream, in which case we assume that the data is
+                // empty.
+                if buffer_len == 0 {
+                    return Ok(0);
+                } else {
+                    return Err(io::ErrorKind::UnexpectedEof.into());
+                }
+            }
+            n => debug_assert_eq!(n, 1),
+        }
+
+        buffer_len += 1;
+
+        match unsigned_varint::decode::usize(&buffer[..buffer_len]) {
+            Ok((len, _)) => return Ok(len),
+            Err(unsigned_varint::decode::Error::Overflow) => {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "overflow in variable-length integer",
+                ));
+            }
+            Err(_) => {}
         }
     }
 }
 
+/// Reads a length-prefixed message from the given socket.
+///
+/// The `max_size` parameter is the maximum size in bytes of the message that we accept. This is
+/// necessary in order to avoid DoS attacks where the remote sends us a message of several
+/// gigabytes.
+///
+/// > **Note**: Assumes that a variable-length prefix indicates the length of the message. This is
+/// >           compatible with what [`write_length_prefixed`] does.
+async fn read_length_prefixed(
+    socket: &mut (impl AsyncRead + Unpin),
+    max_size: usize,
+) -> io::Result<Vec<u8>> {
+    let len = read_varint(socket).await?;
+    if len > max_size {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            format!("Received data size ({len} bytes) exceeds maximum ({max_size} bytes)"),
+        ));
+    }
+
+    let mut buf = vec![0; len];
+    socket.read_exact(&mut buf).await?;
+
+    Ok(buf)
+}
+
 #[async_trait::async_trait]
-impl RequestResponseCodec for MessageCodec {
-    type Protocol = ReqResProtocol;
+impl request_response::Codec for MessageCodec {
+    type Protocol = StreamProtocol;
     type Request = IncomingRequest;
     type Response = OutgoingResponse;
```

Comment:
Whilst certainly possible, this trait was designed with the idea in mind that you'd use _typed_ messages here and not just (length-prefixed) bytes :)

Reply:
Sending bytes back and forth over a "generic" request-response protocol kind of negates the idea of using protocols for multiplexing because you need another, nested layer that allows you to differentiate, which message got sent to you.

YMMV but it might be easier to have multiple instances of `libp2p-request-response`, all with a different codec and more specific protocols that only sent specific messages back and forth (or perhaps only a single message-type per protocol).